migration/migration.go

```diff
@@ -163,3 +175,93 @@ func recalculateBloomFilters(txn db.Transaction) error {
 		}
 	}
 }
+
+var trieNodeBuckets = map[db.Bucket]struct {
+	seekTo  []byte
+	skipLen int
+}{
+	db.ClassesTrie: {
+		seekTo:  db.ClassesTrie.Key(),
+		skipLen: 1,
+	},
+	db.StateTrie: {
+		seekTo:  db.StateTrie.Key(),
+		skipLen: 1,
+	},
+	db.ContractStorage: {
+		seekTo:  db.ContractStorage.Key(),
+		skipLen: 1 + felt.Bytes,
+	},
+}
+
+type trieNode struct {
+	Value *felt.Felt
+	Left  *bitset.BitSet
+	Right *bitset.BitSet
+}
+
+func changeTrieNodeEncoding(txn db.Transaction) error {
+	var n trieNode
+	var buf bytes.Buffer
+	var updatedNodes uint64
+
+	migrateF := func(it db.Iterator, bucket db.Bucket, seekTo []byte, skipLen int) error {
+		bucketPrefix := bucket.Key()
+		for it.Seek(seekTo); it.Valid(); it.Next() {
+			key := it.Key()
+			if !bytes.HasPrefix(key, bucketPrefix) {
+				delete(trieNodeBuckets, bucket)
+				break
+			}
+
+			if len(key) == skipLen {
+				// this entry is not a TrieNode, it is the root key
+				continue
+			}
+
+			const updatedNodesBatch = 1_000_000
+			if updatedNodes >= updatedNodesBatch {
+				trieNodeBuckets[bucket] = struct {
+					seekTo  []byte
+					skipLen int
+				}{seekTo: key, skipLen: trieNodeBuckets[bucket].skipLen}
+
+				return ErrCallWithNewTransaction
+			}
+
+			v, err := it.Value()
+			if err != nil {
+				return err
+			}
+
+			if err = encoder.Unmarshal(v, &n); err != nil {
+				return err
+			}
+
+			coreNode := trie.Node(n)
```

Comment:
We cant because `trie.Node` implements `UnmarshalBinary`.

During the migration, when unmarshaling from the DB, we need the old behaviour (ie use the internal decoder of CBOR). The problem is CBOR shortcircuits and uses the `UnmarshalBinary` function instead of its own decoder for the types that provide `UnmarshalBinary`. So we need a type identical to `trie.Node` but that does not implement `UnmarshalBinary`.